engine/crates/partial-caching/src/fetching/keys.rs

```diff
@@ -0,0 +1,115 @@
+use std::collections::BTreeSet;
+
+use common_types::auth::ExecutionAuth;
+use engine_value::{ConstValue, Variables};
+use registry_for_cache::{CacheAccessScope, CacheControl};
+
+use super::CachingPlan;
+use crate::query_subset::QuerySubsetDisplay;
+
+pub fn build_cache_keys(
+    plan: &CachingPlan,
+    auth: &ExecutionAuth,
+    headers: &http::HeaderMap,
+    variables: &Variables,
+) -> Vec<Option<String>> {
+    plan.cache_partitions
+        .iter()
+        .map(|(cache_control, query_subset)| {
+            // If we can't figure out the scope for a cache_control entry we just skip it.
+            // This will force it to always be fetched from the cache.
+            let scopes = cache_scopes(cache_control, auth, headers)?;
+
+            let serializer = CacheKeySerializer {
+                query: query_subset.as_display(&plan.document),
+                scopes,
+                variables: variables_required(query_subset, &plan.document, variables),
+            };
+
+            match serializer.build_key_string() {
+                Ok(key) => Some(key),
+                Err(err) => {
+                    tracing::error!("Could not build key string for {cache_control:?}: {err:?}");
+                    None
+                }
+            }
+        })
+        .collect()
+}
+
+fn cache_scopes<'a>(
+    cache_control: &CacheControl,
+    auth: &'a ExecutionAuth,
+    headers: &http::HeaderMap,
+) -> Option<CacheAccess<'a>> {
+    let Some(scopes) = &cache_control.access_scopes else {
+        return Some(CacheAccess::Default(auth));
+    };
+
+    let actual_scopes = scopes
+        .iter()
+        .filter_map(|scope| match scope {
+            CacheAccessScope::Public | CacheAccessScope::ApiKey => Some(auth.global_ops().to_string()),
+            CacheAccessScope::Jwt { claim } => auth.as_token().and_then(|token| token.get_claim(claim)),
+            CacheAccessScope::Header { header: header_name } => headers
+                .get(header_name)
+                .and_then(|header| Some(header.to_str().ok()?.to_string())),
+        })
+        .collect::<BTreeSet<_>>();
```

Comment:
Why`BtreeSet`? Cache scope ordering matters afaik

Reply:
Not sure I follow - in what way does it matter?

Reply:
if you have cache scopes: `Jwt(key1), Header(key2)` resulting in the values `Some(X), None` you could also send `None, Some(X)` and have the same cache scope. That's not what I would expect at least